# Using Numpy ndArrays

In [19]:
import os
os.listdir()

['.ipynb_checkpoints', '1.1 smallgray.png.png', 'Numpy and ndArray.ipynb']

Below is how a 2 dimensional array would look in python. This can be used to save an image. **However this is an ineffient way to do so!**

In [49]:
lst = [[12,23,23,3,4,5],[12,23,23,3,4,5],[12,23,23,3,4,5]]

In [11]:
type(lst)

list

#### Rather we use a python library called Numpy.

In [52]:
import numpy as np

In [53]:
n = np.arange(27)

In [54]:
n

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26])

This array is not a python list. **See Below**

In [55]:
type(n)

numpy.ndarray

###### We can easily conver this to a ndarray using the reshape(,) function

In [5]:
n.reshape(3,9)

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23, 24, 25, 26]])

In [6]:
n

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26])

In [7]:
n.reshape(3,3,3)

array([[[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8]],

       [[ 9, 10, 11],
        [12, 13, 14],
        [15, 16, 17]],

       [[18, 19, 20],
        [21, 22, 23],
        [24, 25, 26]]])

In [12]:
type(n)

numpy.ndarray

In [13]:
type(n.reshape(3,3,3))

numpy.ndarray

## Using OpenCV with Numpy

In [2]:
import cv2

cv2 is the OpenCV library. Install it using pip if you dont have it installed.  
*OpenCV uses Numpy to store the images that it reads.*

**Now we are going to open a png file stored locally**  

**We use the `cv2.imread(<filename>,<mode>)` function for reading images in OpenCV.**  
Here `filename` is a string and `mode` is `0` or `1` 

In [6]:
img_rd = cv2.imread("1.1 smallgray.png.png",0)

A png file has got three layers blue, green and red.  
*mode 0 means saving the gray scale* (Creates a 2d Numpy Array)  
*mode 1 means saving the three layers:* **Blue, Green,Red** (Creates a 3d Numpy Array)

In [3]:
img_rd2 = cv2.imread("1.1 smallgray.png.png",1)

In [23]:
img_rd


array([[187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147],
       [209, 134, 255,  97, 182]], dtype=uint8)

In [4]:
img_rd2

array([[[187, 187, 187],
        [158, 158, 158],
        [104, 104, 104],
        [121, 121, 121],
        [143, 143, 143]],

       [[198, 198, 198],
        [125, 125, 125],
        [255, 255, 255],
        [255, 255, 255],
        [147, 147, 147]],

       [[209, 209, 209],
        [134, 134, 134],
        [255, 255, 255],
        [ 97,  97,  97],
        [182, 182, 182]]], dtype=uint8)

##### For writing an array to create an image use the `cv2.imwrite(<filename>,<numpy array>)` method

In [8]:
cv2.imwrite('img_small.png',img_rd)

True

### Slicing and Indexing Numpy Arrays

In [9]:
img_rd

array([[187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147],
       [209, 134, 255,  97, 182]], dtype=uint8)

In [10]:
img_rd[:,1:3]

array([[158, 104],
       [125, 255],
       [134, 255]], dtype=uint8)

In [12]:
img_rd[1:2,1:3]

array([[125, 255]], dtype=uint8)

In [13]:
img_rd[2,4]

182

### Traversing Numpy Arrays

In [14]:
for i in img_rd:
    print(i)

[187 158 104 121 143]
[198 125 255 255 147]
[209 134 255  97 182]


**For obtaining the tranverse of the array use:**

In [15]:
img_rd.T

array([[187, 198, 209],
       [158, 125, 134],
       [104, 255, 255],
       [121, 255,  97],
       [143, 147, 182]], dtype=uint8)

In [19]:
img_rd[0]

array([187, 158, 104, 121, 143], dtype=uint8)

In [20]:
img_rd.T[0]

array([187, 198, 209], dtype=uint8)

*Now suppose you wish to print all the elements of the array in a sequential manner.*  
You could use:

In [18]:
lis = []
for i in img_rd.T:
    lis = lis + (list(i))
lis

[187, 198, 209, 158, 125, 134, 104, 255, 255, 121, 255, 97, 143, 147, 182]

**Or you can use the built in function `flat` of Numpy**

In [24]:
for i in img_rd.T.flat:
    print(i)

187
198
209
158
125
134
104
255
255
121
255
97
143
147
182


In [25]:
for i in img_rd.flat:
    print(i)

187
158
104
121
143
198
125
255
255
147
209
134
255
97
182


### Concatenating and Splitting Numpy Arrays:

In [27]:
img_rd

array([[187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147],
       [209, 134, 255,  97, 182]], dtype=uint8)

In [36]:
img_rd2 = img_rd*2
img_rd2

array([[118,  60, 208, 242,  30],
       [140, 250, 254, 254,  38],
       [162,  12, 254, 194, 108]], dtype=uint8)

*Scalar multiplication multiplies each element by that scalar.*  
Here `dtype=uint8` so maximum range is 255, thus `img_rd2[0,0]` becomes 118, and so on for others

In [28]:
import numpy

**For horizontally stacking arrays:**

In [39]:
np1 = numpy.hstack((img_rd,img_rd2))

In [40]:
np1

array([[187, 158, 104, 121, 143, 118,  60, 208, 242,  30],
       [198, 125, 255, 255, 147, 140, 250, 254, 254,  38],
       [209, 134, 255,  97, 182, 162,  12, 254, 194, 108]], dtype=uint8)

In [41]:
np1 = numpy.hstack((img_rd,img_rd2,img_rd))

In [43]:
print(np1)

[[187 158 104 121 143 118  60 208 242  30 187 158 104 121 143]
 [198 125 255 255 147 140 250 254 254  38 198 125 255 255 147]
 [209 134 255  97 182 162  12 254 194 108 209 134 255  97 182]]


**For vertically stacking arrays:**

In [44]:
np1 = numpy.vstack((img_rd,img_rd2))

In [45]:
print(np1)

[[187 158 104 121 143]
 [198 125 255 255 147]
 [209 134 255  97 182]
 [118  60 208 242  30]
 [140 250 254 254  38]
 [162  12 254 194 108]]


**For vertically splitting arrays:**

In [47]:
print(numpy.vsplit(np1,3))

[array([[187, 158, 104, 121, 143],
       [198, 125, 255, 255, 147]], dtype=uint8), array([[209, 134, 255,  97, 182],
       [118,  60, 208, 242,  30]], dtype=uint8), array([[140, 250, 254, 254,  38],
       [162,  12, 254, 194, 108]], dtype=uint8)]


**For vertically horizontally arrays:**

In [48]:
print(numpy.hsplit(np1,5))

[array([[187],
       [198],
       [209],
       [118],
       [140],
       [162]], dtype=uint8), array([[158],
       [125],
       [134],
       [ 60],
       [250],
       [ 12]], dtype=uint8), array([[104],
       [255],
       [255],
       [208],
       [254],
       [254]], dtype=uint8), array([[121],
       [255],
       [ 97],
       [242],
       [254],
       [194]], dtype=uint8), array([[143],
       [147],
       [182],
       [ 30],
       [ 38],
       [108]], dtype=uint8)]


-----

#### &nbsp;&nbsp;&nbsp;&nbsp; -Archisman Pathak(TheThinker01)